sound/soc/intel/boards/sof_sdw_common.h

```diff
@@ -0,0 +1,126 @@
+/* SPDX-License-Identifier: GPL-2.0
+ *  Copyright (c) 2020 Intel Corporation
+ */
+
+/*
+ *  sof_sdw_common.h - prototypes for common helpers
+ */
+
+#ifndef SND_SOC_SOF_SDW_COMMON_H
+#define SND_SOC_SOF_SDW_COMMON_H
+
+#include <linux/acpi.h>
+#include <linux/async.h>
+#include <linux/delay.h>
+#include <linux/device.h>
+#include <linux/gpio.h>
+#include <linux/init.h>
+#include <linux/input.h>
+#include <linux/io.h>
+#include <linux/dmi.h>
+#include <linux/module.h>
+#include <linux/slab.h>
+#include <linux/soundwire/sdw.h>
+#include <linux/soundwire/sdw_type.h>
+#include <sound/jack.h>
+#include <sound/pcm.h>
+#include <sound/pcm_params.h>
+#include <sound/soc.h>
+#include <sound/soc-acpi.h>
```

Comment:
I very much suspect, that most of these headers aren't needed here. This header only uses bit-operations, `struct list_head` and pointers to other ALSA structs, which can be just forward declared. But many headers here really aren't used at all.

Reply:
Yup, and I personally don't like gathering includes to intermediate .h files like this. I prefer that #include is in the file that needs the interface. I.e. this file should only include headers for definitions that are used directly in sof_sdw_common.h and not all headers that users of sof_sdw_common.h need. 

I understand this can save some typing/lines in the c files, but maintenance is really difficult if you lose the link between include and why it is there. If it's all in the same file, then it's easy to maintain.

Reply:
I don't know how to identify what includes are needed for a specific config. so I took the lazy approach of keeping the list intact and include it everywhere. If you have a better solution that doesn't require trial and errors and works with cross-compilation I am all ears.

Reply:
I think it's actually a requirement (or at least a strong preference) to have those and only those headers included in each specific (.h or .c) file which are needed in *that* file. I.e. if a function is used - a header with its prototype should be included. If a struct / type is used *as an object* - the respective header obviously must be included. If only a pointer is used, it's preferred to forward-declare that type. Obviously for macros, enums... Also recursive includes aren't accepted. E.g. even if the code compiles because a header was included in another included header, it doesn't mean it can be omitted in this file. So we actually have to explicitly include all those kernel.h, types.h etc. headers explicitly where their macros or types are used.

As for how to identify - that's the complex part. In principle you begin with no headers, you write your code, and then every time you use something, which isn't "standard C" you include the respective header.

In this respective case it's pretty simple - the file is small. A good starting point should be
```
#include <linux/bits.h>
#include <linux/types.h>
#include <linux/list.h>
#include <sound/soc.h>
```
That should be it, the rest are structs that you can just forward declare.